### PR TITLE
Use a single HTTP cache in each direction

### PR DESCRIPTION
**Pull request by olix0r**
Currently, all of the fallback logic is implemented outside of the
cache, and then each stack must be cached individualy.

This change consolidates all caching in each direction to a single
cache instance.

This is accomplished by coercing more of the stack to be implemented as
a `NewService`, rather than `MakeService`.

`ResultService` is introduced so that `Admit` can operate on
`NewService`.

### CODE REVIEW

File: linkerd/app/inbound/src/lib.rs [hunk 1]
```diff
@@ -248,50 +249,41 @@ impl Config {
...
+                    .check_new_service::<Target, http::Request<_>>()
                     .into_inner(),
             )
+            .check_new_service::<Target, http::Request<http::boxed::Payload>>()
+            .cache(
+                svc::layers().push_on_response(
+                    svc::layers()
+                        .push_failfast(dispatch_timeout)
+                        .push_spawn_buffer_with_idle_timeout(buffer_capacity, cache_max_idle_age)
+                        .push(metrics.stack.layer(stack_labels("logical")))
```
Comment (hawkw) [inline linkerd/app/inbound/src/lib.rs:281]: nit: is this still per-logical?

File: linkerd/stack/src/result.rs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+use std::task::{Context, Poll};
+
+#[derive(Copy, Clone, Debug)]
+pub struct ResultService<S, E>(Inner<S, E>);
+
+#[derive(Copy, Clone, Debug)]
+enum Inner<S, E> {
+    Ok(S),
+    Err(Option<E>),
+}
```
Comment (hawkw) [inline linkerd/stack/src/result.rs:12]: nit, take it or leave it: isn't this just
```suggestion
pub struct ResultService<S, E>(Result<S, E>);
```

File: linkerd/app/inbound/src/lib.rs (hunk 1 above)
Comment (olix0r) [reply linkerd/app/inbound/src/lib.rs:281]: Yeah, the cache holds a service for each logical target (i.e. the apex service in a traffic split).

**Review by hawkw — APPROVED** [review]
1️⃣💸👍 (get it? "single cache" -> "single cash" 😉)

**Comment by olix0r** [comment]
Merged via https://github.com/linkerd/linkerd2-proxy/pull/671

**Pull request closed without merging**